### ISSUE TITLE
Support setting a partial instead of total state

### ISSUE THREAD

**Issue by samreid** [phetsims/states-of-matter#338]
From phetsims/phet-io#1697: There is 1 or more call to phetioStateEngine.stateSetEmitter.addListener. There is now a 2nd argument to the listener, which is a predicate that tests Tandems and returns true if they are in scope. We should check these listeners and see if any of them need to check whether they are in scope of a partial state set.

**Comment by zepumph** [comment]
In this sim, we have the following listener in MultipleParticleModel.js:

```js
 Tandem.PHET_IO_ENABLED && phet.phetio.phetioEngine.phetioStateEngine.stateSetEmitter.addListener( ( state, inScope ) => {
      // make sure that we have the right number of scaled (i.e. non-normalized) atoms
      const numberOfNormalizedMolecules = this.moleculeDataSet.numberOfMolecules;
      const numberOfNonNormalizedMolecules = this.scaledAtoms.length / this.moleculeDataSet.atomsPerMolecule;
      if ( numberOfNormalizedMolecules > numberOfNonNormalizedMolecules ) {
        this.addAtomsForCurrentSubstance( numberOfNormalizedMolecules - numberOfNonNormalizedMolecules );
      }
      else if ( numberOfNonNormalizedMolecules > numberOfNormalizedMolecules ) {
        _.times( ( numberOfNonNormalizedMolecules - numberOfNormalizedMolecules ) * this.moleculeDataSet.atomsPerMolecule, () => {
          this.scaledAtoms.pop();
        } );
      }

      // clear the injection counter - all atoms and molecules should be accounted for at this point
      this.numMoleculesQueuedForInjectionProperty.reset();

      // synchronize the positions of the scaled atoms to the normalized data set
      this.syncAtomPositions();
    } );
```

This feels very screen dependent, and so I think it should be wrapped in the `isInScope` check.

**Comment by zepumph** [comment]
When I did this, it was working very well, and I am feeling good about the change. I would close this issue, but instead I am going to pass it off to @jbphet to take a look for his knowledge gain. Recently PhET-iO state took on the ability to set subsets of a sim, like a screen. Now the `stateSetEmitter` gives a function that allows you to check to see if this should apply to a provided tandem. I used it in MultipleParticleModel.js as seen above. Feel free to close, ask questions, or discuss.

**Comment by zepumph** [comment]
Just kidding. Please ignore, @samreid and I are changing the API as we speak!

**Comment by zepumph** [comment]
Over in https://github.com/phetsims/tandem/issues/193, we altered the api for the stateSetEmitter slightly. Now instead of checking a function, you check against a "scopeTandem" that serves as the root of the tree being state set. 

@jbphet everything in https://github.com/phetsims/states-of-matter/issues/338#issuecomment-676670057 still applies except that the api changed from checking based on a function, to checking based on a Tandem. See https://github.com/phetsims/states-of-matter/commit/ba4746a10c695b35de1cf3a0bcc77db642b20fd4. I tested in studio by customizing some stuff on the first two screens relating to the particles, and then launching. I saw that full state set correctly. Then after playing with it a bit more, I saw that reset worked as expected, setting back to the launched state.

Please note the update to the PhET-iO state API. Feel free to close.

**Comment by kathy-phet** [comment]
We want to understand this new feature.

**Comment by jbphet** [comment]
In #333 I ended up moving the code that was hooked to the `stateSetEmitter` into the `applyState` method so that all of the state setting code was in one place.  This seems to work fine in testing.

I believe that this makes much of the discussion in this issue moot, and it obviates the need for a `isInScope` check.  I'm not 100% certain, though, so I'd like @samreid to review and make sure that this has been handled correctly with respect to phet-io state setting.

@samreid - My attempt to boil it down to a single question - and thereby save you time - is this: If the `stateSetEmitter` is no longer being used, is there any reason to worry about which tandems are and are not in scope?  If the answer is no, I think this can be closed.  If it's more complicated than that, I'll need to work with you to understand why and what should be done about it.

**Comment by zepumph** [comment]
>If the stateSetEmitter is no longer being used, is there any reason to worry about which tandems are and are not in scope? If the answer is no, I think this can be closed. If it's more complicated than that, I'll need to work with you to understand why and what should be done about it.

I was here and read through this, so I'll just answer for the phet-io context. . . .

Your above commit (https://github.com/phetsims/states-of-matter/commit/b78f16320d6a28495714bb70aa9e2a6cad7e7b6d) does a very good job of eliminating the problem, because `applyState` will only be called for that particular instance of the model (screen-specific). 

The question still remains though if this patch made it into the currently released sim, and, if not, if that is an issue. 

Again to reiterate:
1. I love you commit, it is clean and great because it gets rid of a usage of `setStateEmitter.addListener()`.
2. The fact remains that the potentially buggy code is in the published sim: https://github.com/phetsims/states-of-matter/blob/5ca393b5e8801287b6e9c3e337881ea8e82f0d18/js/common/model/MultipleParticleModel.js#L316-L336. This is state at the head of the 1.2 branch. Thus my question remains from https://github.com/phetsims/states-of-matter/issues/338#issuecomment-678533063:

>Does that above, wrapped code in the setStateEmitter cause a problem if it was called on the screen 1 model when the screen 2 reset all button reset, setting only the state for screen 2?

Up to you!

**Comment by jbphet** [comment]
Thanks @zepumph, I understand the concern you're expressing.  The first portion of the code in question won't be executed in the described case, because there will be no discrepancy between the number of non-normalized and normalized particles on the non-active screen.  The last part - the call to `syncAtomPositions` will essentially be a no-op, because it will be synchronizing particles that would have no reason at that point be be unsynchronized.  So, the only part that could potentially cause a problem would be the code that clears the count of molecules that are queued for injection.  I went to the State wrapper in the live version (1.2.0) to see if I could ever create a situation where there were queued particles on one screen and then perform a reset on the other.  I couldn't.  One would have to move really fast.  And, even if the counter had something in it and it got cleared, a subsequent setting of the state would clear up the problem, since any missing particles would be added as part of the state.  I also just messed around a bunch with setting up state on screens 1 & 2 and switching back and forth between them and pressing reset while running in the state wrapper, and I didn't see anything odd.

Bottom line (literally): I don't think there is any need to do a maintenance release with this patch.  Closing.
